src/services/refactors/inlineVariable.ts

```diff
@@ -0,0 +1,194 @@
+        if (!info) {
+            return emptyArray;
+        }
+
+        if (refactor.isRefactorErrorInfo(info) && preferences.provideRefactorNotApplicableReason) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [{
+                    ...inlineVariableAction,
+                    notApplicableReason: info.error
+                }]
+            }];
+        }
+
+        if (!refactor.isRefactorErrorInfo(info)) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [inlineVariableAction]
+            }];
+        }
+
+        return emptyArray;
+    },
+
+    getEditsForAction(context, actionName) {
+        Debug.assert(actionName === refactorName, "Unexpected refactor invoked");
+
+        const {
+            cancellationToken,
+            file,
+            program,
+            startPosition,
+        } = context;
+
+        const info = getInliningInfo(file, startPosition, /*tryWithReferenceToken*/ true, program, cancellationToken);
+
+        if (!info || refactor.isRefactorErrorInfo(info)) {
+            return undefined;
+        }
+
+        const { references, declaration, replacement } = info;
+        const edits = textChanges.ChangeTracker.with(context, tracker => {
+            for (const node of references) {
+                tracker.replaceNode(file, node, getReplacementExpression(node, replacement));
+            }
+            tracker.delete(file, declaration);
+        });
+
+        return { edits };
+    }
+});
+
+function getInliningInfo(file: SourceFile, startPosition: number, tryWithReferenceToken: boolean, program: Program, cancellationToken: CancellationToken = NoopCancellationToken): InliningInfo | RefactorErrorInfo | undefined {
+    const token = getTokenAtPosition(file, startPosition);
+    const parent = token.parent;
+
+    // If the node is a variable declaration, make sure it's not in a catch clause or for-loop
+    // and that it has a value.
+    if (isInitializedVariable(parent) && isVariableDeclarationInVariableStatement(parent)) {
+        // Find all references to the variable.
+        const name = parent.name;
+        const referenceEntries = getReferenceEntriesForNode(name.pos, name, program, program.getSourceFiles(), cancellationToken);
+        if (!referenceEntries) {
+            return undefined;
+        }
+        const referenceNodes = getReferenceNodes(referenceEntries, name);
+
+        return referenceNodes && { references: referenceNodes, declaration: parent, replacement: parent.initializer };
+    }
+
+    if (tryWithReferenceToken && isIdentifier(token)) {
+        // Try finding the declaration and nodes to replace via the reference token.
+        const referencedSymbols = FindAllReferences.Core.getReferencedSymbolsForNode(token.pos, token, program, program.getSourceFiles(), cancellationToken);

```

Comment:
It looks like this may find references in other files. Is that intended? If so, it looks like it needs to be tested. If the symbol is getting imported in other files and then getting inlined in those files, we’ll need to delete the import in those files as well. I think that would be a neat feature, but maybe beyond the scope of what you planned. A simpler approach to start with would be to disable the refactor on exported symbols&sup1; and only search local references (`FindAllReferences.Core.eachSymbolReferenceInFile`).

&sup1; Note that a symbol can be exported somewhere else in the file, so it’s not enough to check for an `export` modifier:

```ts
const x = 1 + 2;

// ... some inlinable references to `x`, maybe

export { x };
```

Reply:
I previously tested this and exported symbols won't be renamed because [isWriteAccessForReference](https://github.com/microsoft/TypeScript/blob/b58457b32be2008b15f4fbe1ee0e4a0fe38352bb/src/services/findAllReferences.ts#L856) returns true with imports and exports (because of [this](https://github.com/microsoft/TypeScript/blob/b58457b32be2008b15f4fbe1ee0e4a0fe38352bb/src/services/findAllReferences.ts#L881)). I'm only providing the refactoring if all references are reads.

I will add unit tests for this though.

Reply:
In that case, you may want to find a way to avoid searching for references in all files, since that’s very expensive and you won’t do anything with them except bail. `FindAllReferences.Core.eachSymbolReferenceInFile` only looks in one file, but that may break the way you’re currently relying on determining if something is imported.

Reply:
So I tried using `eachSymbolReferenceInFile` but the issue is that it would only work when I have access to the declaration node, but I want to also handle the case where I'm starting from a reference and want the declaration back. Also, the nice thing about `getReferencedSymbolsForNode` is that it will explicitly separate the declaration from the references, which can get messy with `eachSymbolReferenceInFile`.

Could I avoid the expensive operation by just limiting the `sourceFiles` parameter to the file in the refactor context? I think this will limit the search to that set.

Reply:
Maybe? I’m not too familiar with the FAR code. Getting the declaration is no problem though, that’s just `checker.resolveName`.

Reply:
@weswigham Do you know if limiting `sourceFiles` will address @andrewbranch's perf concern with `getReferencedSymbolsForNode`?

Reply:
Nvm I ended up following @andrewbranch's suggestion.